odl/contrib/shearlab/shearlab_operator.py

```diff
@@ -0,0 +1,312 @@
+
+
+# Load Shearlab
+j = load_Shearlab()
+
+
+# Function to load images with certain size
+def load_image(name, n, m=None, gpu=0, square=0):
+    if m is None:
+        m = n
+    command = 'Shearlab.load_image(' + '"' + name + '"' + ',' + str(n) + ','
+    command = command + str(m) + ',' + str(gpu) + ',' + str(square) + ')'
+    return j.eval(command)
+
+
+# Function to plot images
+def imageplot(f, str='', sbpt=[]):
+    """
+        Use nearest neighbor interpolation for the display.
+    """
+    if sbpt != []:
+        plt.subplot(sbpt[0], sbpt[1], sbpt[2])
+    imgplot = plt.imshow(f, interpolation='nearest')
+    imgplot.set_cmap('gray')
+    pylab.axis('off')
+    if str != '':
+        plt.title(str)
+
+
+# Class of shearlet system in 2D
+class Shearletsystem2D:
+    def __init__(self, shearlets, size, shearLevels, full, nShearlets,
+                 shearletIdxs, dualFrameWeights, RMS, isComplex):
+        self.shearlets = shearlets
+        self.size = size
+        self.shearLevels = shearLevels
+        self.full = full
+        self.nShearlets = nShearlets
+        self.shearletIdxs = shearletIdxs
+        self.dualFrameWeights = dualFrameWeights
+        self.RMS = RMS
+        self.isComplex = isComplex
+
+
+# Function to generate de 2D system
+def getshearletsystem2D(rows, cols, nScales, shearLevels=None,
+                        full=0,
+                        directionalFilter='Shearlab.
+                        filt_gen("directional_shearlet")',
+                        quadratureMirrorFilter='Shearlab.
+                        filt_gen("scaling_shearlet")'):
+    if shearLevels is None:
+        shearLevels = [float(ceil(i / 2)) for i in range(1, nScales + 1)]
+    j.eval('rows=' + str(rows))
+    j.eval('cols=' + str(cols))
+    j.eval('nScales=' + str(nScales))
+    j.eval('shearLevels=' + str(shearLevels))
+    j.eval('full=' + str(full))
+    j.eval('directionalFilter=' + directionalFilter)
+    j.eval('quadratureMirrorFilter=' + quadratureMirrorFilter)
+    j.eval('shearletSystem=Shearlab.getshearletsystem2D(rows,
+           cols, nScales, shearLevels, full, directionalFilter,
+           quadratureMirrorFilter) ')
+    shearlets = j.eval('shearletSystem.shearlets')
+    size = j.eval('shearletSystem.size')
+    shearLevels = j.eval('shearletSystem.shearLevels')
+    full = j.eval('shearletSystem.full')
+    nShearlets = j.eval('shearletSystem.nShearlets')
+    shearletIdxs = j.eval('shearletSystem.shearletIdxs')
+    dualFrameWeights = j.eval('shearletSystem.dualFrameWeights')
+    RMS = j.eval('shearletSystem.RMS')
+    isComplex = j.eval('shearletSystem.isComplex')
+    return Shearletsystem2D(shearlets, size, shearLevels, full, nShearlets,
+                            shearletIdxs, dualFrameWeights, RMS, isComplex)
+
+
+# Shearlet Decomposition function
+def sheardec2D(X, shearletSystem):
+    coeffs = np.zeros(shearletSystem.shearlets.shape, dtype=np.complex_)
+    Xfreq = fftshift(fft2(ifftshift(X)))
```

Comment:
I will disagree on you with the heavy work is done in Python. Actually the heavy work in the Shearlet transform implementation is done in the generation of the corresponding system which in this case is in julia, the decomposition takes nothing in comparison with the generation. Of course there is a implementation of all that function in julia, since my library Shearlab.jl is fully written in julia (check the corresponding [implemenations](https://github.com/arsenal9971/Shearlab.jl/blob/master/src/2D/shearletdecrec2D.jl)).

The reason why I tried to implement this in python is more related on how pyjulia works, the object called Shearlet system is a special type in julia created by me. When trying to import this from julia to python I am not sure how this will work. I also don't know well how to import julia objects into the python enviroment, i know julia arrays can be imported as numpy arrays, I used this fact in the implementation. I think I have a good idea to do this, just give me between today and tomorrow and I will find the way to do also the decomposition and everything in julia. I also will run some benchmarks which case is better.

Reply:
Okay, fair enough. There you see my ignorance of what happens under the hood :-)
What's your rough estimate on the distribution of the numerical effort between generating the system and performing the decomposition? I'm asking to see if it may still be worth to improve efficiency on the Python side.

Reply:
I runned some [benchmarks](https://github.com/arsenal9971/odl/blob/master/odl/contrib/shearlab/examples/Benchmarks.ipynb), and you can see that the system generation takes most of the time, and is way faster in my version than pyshearlab (around 9 seconds vs 44). You can also see that the decomposition and reconstruction doesnt take so much time (both around 2 seconds), and running pure julia code it just improve around .5 seconds.

 It will also take time if I want to do everything in julia to stream the numpy array that we want to analyse to a julia array, and it will also take more memory, therefore I dont think is worthwhile to do a pure julia implementation of this. I also think that adding a todo to use pyFFTW is a good idea.